batch_job/src/batch_job_k8s.c

```diff
@@ -0,0 +1,463 @@
+			debug(D_BATCH, "Successfully delete pods %s", pod_id);
+		}
+		
+		free(pod_id);
+		return 0;
+
+    } else if(pid < 0) {
+
+		fatal("couldn't create new process: %s\n", strerror(errno));
+
+    } else {
+
+		char *cmd = string_format("kubectl delete pods %s", pod_id);
+
+		execlp("/bin/bash", "bash", "-c", cmd, (char *) 0);
+		_exit(errno);
+
+    }
+
+    return 1;	
+
+}
+
+static void batch_job_k8s_handle_complete_task (char *pod_id, int job_id,
+		k8s_job_info *curr_k8s_job_info, int exited_normally, 
+		struct batch_job_info *info_out, struct list *running_pod_lst,
+		struct batch_queue *q)
+{
+	struct batch_job_info *info;
+	int timeout = 5;
+	debug(D_BATCH, "%d is failed to execute.", job_id);
+	info = itable_remove(q->job_table, job_id);
+	info->exited_normally = exited_normally;
+	// TODO get exact exit_code
+	if (exited_normally == 0) {
+		info->exit_code = 1;
+	}
+	memcpy(info_out, info, sizeof(*info));
+	free(info);
+
+	list_remove(running_pod_lst, pod_id);
+	batch_job_k8s_remove(q, job_id);
+
+	if(curr_k8s_job_info->is_running == 0) {
+		process_wait(timeout);
+	} else {
+		process_wait(timeout);
+		process_wait(timeout);
+	}
+
+}
+
+static batch_job_id_t batch_job_k8s_wait (struct batch_queue * q, 
+		struct batch_job_info * info_out, time_t stoptime)
+{
+	/*
+	 * There are 5 states for a k8s job
+	 * 1. pod_created
+	 * 2. inps_transferred
+     * 3. exec_success/exec_failed
+     * 4. oups_transferred
+	 * 5. job_done 
+	 */
+
+	struct list *running_pod_lst = NULL;
+
+	while(1) {
+	
+		// generate the list of running pod	
+		if(running_pod_lst != NULL) {
+			list_free(running_pod_lst);	
+			list_delete(running_pod_lst);
+		}
+		
+		running_pod_lst = list_create();	
+		
+		char *cmd = string_format("kubectl get pods -l app=%s | awk \'{if (NR != 1) {print $1\" \"$3}}\' 2>&1 ", 
+				mf_uuid->str);
+		FILE *cmd_fp;
+		char pod_info[128];
+		cmd_fp = popen(cmd, "r");
+
+		while(fgets(pod_info, sizeof(pod_info)-1, cmd_fp) != NULL) {
+			char *pch, *pod_id, *pod_state;
+			pch = strtok(pod_info, " \t");
+			pod_id = xxstrdup(pch);
+			pod_state = strtok(NULL, " \t");
+
+			int i = 0;
+			while(pod_state[i] != '\n' && pod_state[i] != EOF)
+				i ++;
+			if(pod_state[i] == '\n' || pod_state[i] == EOF) 
+				pod_state[i] = '\0';
+			
+
+			if(strcmp(pod_state, "Running") == 0) {
+				list_push_tail(running_pod_lst, (void *) pod_id);
+			} 
+
+			if(strcmp(pod_state, "Failed") == 0) {
+				char *curr_job_id = xxstrdup(strrchr(pod_id, '-') + 1);	
+				int curr_job_id_int = atoi(curr_job_id);
+				k8s_job_info *curr_k8s_job_info = itable_lookup(k8s_job_info_table, curr_job_id_int);
+				batch_job_k8s_handle_complete_task(pod_id, curr_job_id_int, 
+						curr_k8s_job_info, 1, info_out, running_pod_lst, q);
+				return curr_job_id_int;
+			}
+		}
+		pclose(cmd_fp);
+
+		// iterate the running pods
+		char *curr_pod_id;
+		list_first_item(running_pod_lst);	
+		while((curr_pod_id = (char *)list_next_item(running_pod_lst))) {
+			
```

Comment:
Factor the interior of this loop into a separate function get_pod_status or similar